### PR TITLE
Use OPENSSL_free instead of free to avoid instance crash

### PR DESCRIPTION
**Pull request by haidong**
## Description
When MariaDB is built with OpenSSL forks other than wolfSSL, a user defined like this:
`grant select on test.* to user1@localhost require SUBJECT "...";`
may crash the server during connection.

The crash is caused by the usage of `free` for memory cleanup. OpenSSL uses [**OPENSSL_xxx** API for memory management](https://www.openssl.org/docs/man1.1.1/man3/OPENSSL_malloc.html). For allocation, `OPENSSL_malloc` is used. To free it, `OPENSSL_free` should be called, _not_ `free`.

We've been lucky so far with `OpenSSL` and `wolfSSL`, because their implementation allow it. Other forks such as [AWS-LC](https://github.com/awslabs/aws-lc) may not be so forgiving.

There have been numerous [OpenSSL forks](https://en.wikipedia.org/wiki/OpenSSL#Forks). Due to [FIPS](https://en.wikipedia.org/wiki/FIPS_140-2) and other related regulatory requirements, MariaDB will be built using them. This fix will increase MariaDB's adaptability by using more compliant and generally accepted API.

## How can this PR be tested?
All existing test cases should still be passing in Buildbot as well as GitLab-CI.

In particular, the existing test case `openssl_1` led to the discovery of this bug, and will provide excellent coverage for this PR.

## Basing the PR against the correct MariaDB version
- [x] This is a bug fix and the PR is based against the long-term support release, which is 10.3 as of this writing.

## Backward compatibility
This bug fix should be fully backward-compatible

## Copyright
All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

### CODE REVIEW

**Review by dlenski — APPROVED** [review]

**Comment by dlenski** [comment]
Nice. It might be possible to automate checks for other cases where mismatched alloc/free functions are used. From [Valgrind Memcheck docs](https://valgrind.org/docs/manual/mc-manual.html#mc-manual.rudefn):

image

By augmenting the alloc/free functions' source code with [the appropriate Valgrind macros](https://developers.redhat.com/articles/2022/03/23/use-valgrind-memcheck-custom-memory-manager#tracking_with_block_macros) (`VALGRIND_MALLOCLIKE_BLOCK` and `VALGRIND_FREELIKE_BLOCK`), it would then be possible to use memcheck to find any/all cases where the wrong alloc/free pair is used.

**Review by dlenski — APPROVED** [review]

**Comment by haidong** [comment]
Thanks @dlenski !

> By augmenting the alloc/free functions' source code with [the appropriate Valgrind macros](https://developers.redhat.com/articles/2022/03/23/use-valgrind-memcheck-custom-memory-manager#tracking_with_block_macros) (VALGRIND_MALLOCLIKE_BLOCK and VALGRIND_FREELIKE_BLOCK), it would then be possible to use memcheck to find any/all cases where the wrong alloc/free pair is used.

That seems like a great idea and potentially very useful!

**Comment by grooverdan** [comment]
The openssl code maps the `X509_NAME_oneline` return ptr with a null parameter to be released by `OPENSSL_free` so works for me:
* [ref 1](https://github.com/openssl/openssl/blob/79c8dcf3985a7b75eac8e53eb8652728af6c5d3d/crypto/x509/x509_cmp.c#L47-L68)
* [ref 2](https://github.com/openssl/openssl/blob/a73bdc24e14760413a65d478d7c88356b4b95bb5/crypto/x509/x_name.c#L508-L512)
* [ref 3](https://github.com/openssl/openssl/blob/e1e93f7a07dfc7a8dddd4ddbb79d1d9bc9760d32/crypto/cmp/cmp_vfy.c#L157-L162)

**Comment by dlenski** [comment]
> Would be nice if their `X509_NAME_oneline` man page included such detail.

Very much agree.

The man page for [`openssl_free`, `openssl_malloc`](https://www.openssl.org/docs/man3.0/man3/OPENSSL_free.html) is quite long and hard to follow.

I can't find _any_ explicit mention in the OpenSSL docs that _all_ of the newly-allocated buffers returned by the library need `openssl_free`, but perhaps someone else knows where that lives.

**Comment by grooverdan** [comment]
And `X509_free` for certs, however that's different.

**Comment by grooverdan** [comment]
Thanks @haidong and @dlenski
